Error replicating from source cluster running Hive 1.2.1 to replica cluster running Hive 2.3.2

The following error occurs when running Circus Train on EMR with Hive 2.3.2 libraries performing a replication where the source is running Hive 1.2.1 and replica is running 2.3.2. The issue appears to be a Thrift backwards incompatability issue between a newer Hive client and an older Hive server. Here is an example failing log output (with certain host names redacted):

> Using /home/hadoop/circus-train-11.1.1-SNAPSHOT
> SLF4J: Class path contains multiple SLF4J bindings.
> SLF4J: Found binding in [jar:file:/usr/lib/hadoop/lib/slf4j-log4j12-1.7.10.jar!/org/slf4j/impl/StaticLoggerBinder.class]
> SLF4J: Found binding in [jar:file:/home/hadoop/circus-train-11.1.1-SNAPSHOT/lib/slf4j-log4j12-1.7.21.jar!/org/slf4j/impl/StaticLoggerBinder.class]
> SLF4J: See http://www.slf4j.org/codes.html#multiple_bindings for an explanation.
> SLF4J: Actual binding is of type [org.slf4j.impl.Log4jLoggerFactory]
> 18/02/12 11:22:53 INFO circustrain.CircusTrain: Maven-ArtifactId=circus-train-core; Built-By=ddelcastilloperez; Build-DateTime=2018-02-12T11:20:19Z; Created-By=Apache Maven 3.5.2; Build-Jdk=1.8.0_152; Build-Version=11.1.1-SNAPSHOT; Manifest-Version=1.0; Maven-GroupId=com.hotels;
> 18/02/12 11:22:54 INFO util.Version: HV000001: Hibernate Validator 5.2.4.Final
>          ee@@@@@@@@@@@@@@@
>        e@@@@@@@@@@@@@@@
>       @@@"     .-.----.      ___ _                   ___ _                   _
>      @@"___   / o )    \    / __\ |__   ___   ___   / __\ |__   ___   ___   / \
>     II__[w]   || ´ __  |'  / /  | '_ \ / _ \ / _ \ / /  | '_ \ / _ \ / _ \ /  /
>    {======|_ '- |||  |||  / /___| | | | (_) | (_) / /___| | | | (_) | (_) /\_/
>   /oO--000'"`--OO----OO-' \____/|_| |_|\___/ \___/\____/|_| |_|\___/ \___/\/
> 18/02/12 11:22:54 INFO circustrain.CircusTrain: Starting CircusTrain on ip-10-29-171-135.us-west-2.compute.internal with PID 5196 (/home/hadoop/circus-train-11.1.1-SNAPSHOT/lib/circus-train-all-latest.jar started by hadoop in /home/hadoop)
> 18/02/12 11:22:54 INFO circustrain.CircusTrain: The following profiles are active: replication,housekeeping
> 18/02/12 11:22:54 INFO annotation.AnnotationConfigApplicationContext: Refreshing org.springframework.context.annotation.AnnotationConfigApplicationContext@fb5aeed: startup date [Mon Feb 12 11:22:54 UTC 2018]; root of context hierarchy
> 18/02/12 11:22:57 WARN annotation.ConfigurationClassEnhancer: @Bean method EnableEncryptablePropertySourcesConfiguration.enableEncryptablePropertySourcesPostProcessor is non-static and returns an object assignable to Spring's BeanFactoryPostProcessor interface. This will result in a failure to process annotations such as @Autowired, @Resource and @PostConstruct within the method's declaring @Configuration class. Add the 'static' modifier to this method to avoid these container lifecycle issues; see @Bean javadoc for complete details.
> 18/02/12 11:22:57 INFO jasyptspringboot.EnableEncryptablePropertySourcesPostProcessor: Post-processing PropertySource instances
> 18/02/12 11:22:57 INFO jasyptspringboot.EnableEncryptablePropertySourcesPostProcessor: Converting PropertySource commandLineArgs[org.springframework.core.env.SimpleCommandLinePropertySource] to EncryptableEnumerablePropertySourceWrapper
> 18/02/12 11:22:57 INFO jasyptspringboot.EnableEncryptablePropertySourcesPostProcessor: Converting PropertySource systemProperties[org.springframework.core.env.MapPropertySource] to EncryptableMapPropertySourceWrapper
> 18/02/12 11:22:57 INFO jasyptspringboot.EnableEncryptablePropertySourcesPostProcessor: Converting PropertySource systemEnvironment[org.springframework.core.env.SystemEnvironmentPropertySource] to EncryptableMapPropertySourceWrapper
> 18/02/12 11:22:57 INFO jasyptspringboot.EnableEncryptablePropertySourcesPostProcessor: Converting PropertySource random[org.springframework.boot.context.config.RandomValuePropertySource] to EncryptablePropertySourceWrapper
> 18/02/12 11:22:57 INFO jasyptspringboot.EnableEncryptablePropertySourcesPostProcessor: Converting PropertySource applicationConfig: [file:/home/hadoop/BDP-1151.yml][org.springframework.core.env.MapPropertySource] to EncryptableMapPropertySourceWrapper
> 18/02/12 11:22:57 INFO jasyptspringboot.EnableEncryptablePropertySourcesPostProcessor: Converting PropertySource defaultProperties[org.springframework.core.env.MapPropertySource] to EncryptableMapPropertySourceWrapper
> 18/02/12 11:22:57 INFO annotation.AutowiredAnnotationBeanPostProcessor: JSR-330 'javax.inject.Inject' annotation found and supported for autowiring
> 18/02/12 11:22:57 INFO support.PostProcessorRegistrationDelegate$BeanPostProcessorChecker: Bean 'org.springframework.transaction.annotation.ProxyTransactionManagementConfiguration' of type [class org.springframework.transaction.annotation.ProxyTransactionManagementConfiguration$$EnhancerBySpringCGLIB$$85b58278] is not eligible for getting processed by all BeanPostProcessors (for example: not eligible for auto-proxying)
> 18/02/12 11:22:57 INFO support.PostProcessorRegistrationDelegate$BeanPostProcessorChecker: Bean 'bindS3AFileSystem' of type [class com.hotels.bdp.circustrain.aws.BindS3AFileSystem] is not eligible for getting processed by all BeanPostProcessors (for example: not eligible for auto-proxying)
> 18/02/12 11:22:57 INFO support.PostProcessorRegistrationDelegate$BeanPostProcessorChecker: Bean 's3CredentialsUtils' of type [class com.hotels.bdp.circustrain.aws.S3CredentialsUtils] is not eligible for getting processed by all BeanPostProcessors (for example: not eligible for auto-proxying)
> 18/02/12 11:22:57 INFO support.PostProcessorRegistrationDelegate$BeanPostProcessorChecker: Bean 'housekeepingConfiguration' of type [class com.hotels.housekeeping.HousekeepingConfiguration$$EnhancerBySpringCGLIB$$fd62bc27] is not eligible for getting processed by all BeanPostProcessors (for example: not eligible for auto-proxying)
> 18/02/12 11:22:58 INFO support.PostProcessorRegistrationDelegate$BeanPostProcessorChecker: Bean 'stringToDurationConverter' of type [class com.hotels.housekeeping.converter.StringToDurationConverter] is not eligible for getting processed by all BeanPostProcessors (for example: not eligible for auto-proxying)
> 18/02/12 11:22:58 INFO support.PostProcessorRegistrationDelegate$BeanPostProcessorChecker: Bean 'GCPSecurity' of type [class com.hotels.bdp.circustrain.gcp.context.GCPSecurity$$EnhancerBySpringCGLIB$$3b0e0b7b] is not eligible for getting processed by all BeanPostProcessors (for example: not eligible for auto-proxying)
> 18/02/12 11:22:58 INFO hikari.HikariDataSource: HikariPool-1 - Starting...
> 18/02/12 11:22:59 INFO hikari.HikariDataSource: HikariPool-1 - Start completed.
> 18/02/12 11:22:59 INFO init.ScriptUtils: Executing SQL script from URL [jar:file:/home/hadoop/circus-train-11.1.1-SNAPSHOT/lib/circus-train-all-latest.jar!/schema.sql]
> 18/02/12 11:22:59 INFO init.ScriptUtils: Executed SQL script from URL [jar:file:/home/hadoop/circus-train-11.1.1-SNAPSHOT/lib/circus-train-all-latest.jar!/schema.sql] in 5 ms.
> 18/02/12 11:22:59 INFO init.ScriptUtils: Executing SQL script from URL [file:/tmp/hadoop-unjar5929440329897995965/schema.sql]
> 18/02/12 11:22:59 INFO init.ScriptUtils: Executed SQL script from URL [file:/tmp/hadoop-unjar5929440329897995965/schema.sql] in 0 ms.
> 18/02/12 11:22:59 INFO jpa.LocalContainerEntityManagerFactoryBean: Building JPA container EntityManagerFactory for persistence unit 'default'
> 18/02/12 11:22:59 INFO util.LogHelper: HHH000204: Processing PersistenceUnitInfo [
>     name: default
>     ...]
> 18/02/12 11:22:59 INFO hibernate.Version: HHH000412: Hibernate Core {4.3.11.Final}
> 18/02/12 11:22:59 INFO cfg.Environment: HHH000206: hibernate.properties not found
> 18/02/12 11:22:59 INFO cfg.Environment: HHH000021: Bytecode provider name : javassist
> 18/02/12 11:23:01 INFO common.Version: HCANN000001: Hibernate Commons Annotations {4.0.5.Final}
> 18/02/12 11:23:01 INFO dialect.Dialect: HHH000400: Using dialect: org.hibernate.dialect.H2Dialect
> 18/02/12 11:23:01 INFO ast.ASTQueryTranslatorFactory: HHH000397: Using ASTQueryTranslatorFactory
> 18/02/12 11:23:01 INFO hbm2ddl.SchemaUpdate: HHH000228: Running hbm2ddl schema update
> 18/02/12 11:23:01 INFO hbm2ddl.SchemaUpdate: HHH000102: Fetching database metadata
> 18/02/12 11:23:01 INFO hbm2ddl.SchemaUpdate: HHH000396: Updating schema
> 18/02/12 11:23:01 INFO hbm2ddl.TableMetadata: HHH000261: Table found: HOUSEKEEPING.CIRCUS_TRAIN.AUDIT_REVISION
> 18/02/12 11:23:01 INFO hbm2ddl.TableMetadata: HHH000037: Columns: [id, timestamp]
> 18/02/12 11:23:01 INFO hbm2ddl.TableMetadata: HHH000108: Foreign keys: []
> 18/02/12 11:23:01 INFO hbm2ddl.TableMetadata: HHH000126: Indexes: [primary_key_5]
> 18/02/12 11:23:01 INFO hbm2ddl.TableMetadata: HHH000261: Table found: HOUSEKEEPING.CIRCUS_TRAIN.LEGACY_REPLICA_PATH
> 18/02/12 11:23:01 INFO hbm2ddl.TableMetadata: HHH000037: Columns: [path, event_id, creation_timestamp, path_event_id, id]
> 18/02/12 11:23:01 INFO hbm2ddl.TableMetadata: HHH000108: Foreign keys: []
> 18/02/12 11:23:01 INFO hbm2ddl.TableMetadata: HHH000126: Indexes: [primary_key_1, uk_tlauv1txmi2vux3fqp1y1ueav_index_1]
> 18/02/12 11:23:01 INFO hbm2ddl.SchemaUpdate: HHH000232: Schema update complete
> 18/02/12 11:23:01 INFO circustrain.CircusTrain: Maven-ArtifactId=circus-train-core; Built-By=ddelcastilloperez; Build-DateTime=2018-02-12T11:20:19Z; Created-By=Apache Maven 3.5.2; Build-Jdk=1.8.0_152; Build-Version=11.1.1-SNAPSHOT; Manifest-Version=1.0; Maven-GroupId=com.hotels;
> 18/02/12 11:23:02 INFO context.CommonBeans: No Hadoop site XML is defined for catalog karmalab-dev.
> 18/02/12 11:23:02 INFO conf.HiveConf: Found configuration file file:/etc/hive/conf.dist/hive-site.xml
> 18/02/12 11:23:02 INFO context.CommonBeans: No Hadoop site XML is defined for catalog aws-data-lab.
> 18/02/12 11:23:03 WARN impl.FileSystemHousekeepingService: class com.hotels.housekeeping.service.impl.FileSystemHousekeepingService.fixIncompleteRecord(LegacyReplicaPath) should be removed in future.
> 18/02/12 11:23:04 INFO annotation.AnnotationMBeanExporter: Registering beans for JMX exposure on startup
> 18/02/12 11:23:04 INFO annotation.AnnotationMBeanExporter: Bean with name 'housekeepingDataSource' has been autodetected for JMX exposure
> 18/02/12 11:23:04 INFO annotation.AnnotationMBeanExporter: Located MBean 'housekeepingDataSource': registering with JMX server as MBean [com.zaxxer.hikari:name=housekeepingDataSource,type=HikariDataSource]
> 18/02/12 11:23:04 INFO core.Locomotive: 1 tables to replicate.
> 18/02/12 11:23:04 INFO core.Locomotive: Replicating karmalab-dev:ddelcastilloperez.access_log to aws-data-lab:bdp.access_log_dev replication mode 'FULL'.
> 18/02/12 11:23:04 INFO hive.metastore: Trying to connect to metastore with URI thrift://cheledwhdd004.karmalab.net:9083
> 18/02/12 11:23:04 INFO hive.metastore: Opened a connection to metastore, current connections: 1
> 18/02/12 11:23:04 INFO hive.metastore: Connected to metastore.
> 18/02/12 11:23:04 INFO hive.metastore: Closed a connection to metastore, current connections: 0
> 18/02/12 11:23:04 INFO hive.metastore: Trying to connect to metastore with URI thrift://replica:9083
> 18/02/12 11:23:04 INFO hive.metastore: Opened a connection to metastore, current connections: 1
> 18/02/12 11:23:04 INFO hive.metastore: Connected to metastore.
> 18/02/12 11:23:04 INFO hive.metastore: Closed a connection to metastore, current connections: 0
> 18/02/12 11:23:04 INFO source.Source: Retrieving table metadata for 'ddelcastilloperez.access_log'
> 18/02/12 11:23:04 INFO hive.metastore: Trying to connect to metastore with URI thrift://cheledwhdd004.karmalab.net:9083
> 18/02/12 11:23:04 INFO hive.metastore: Opened a connection to metastore, current connections: 1
> 18/02/12 11:23:05 INFO hive.metastore: Connected to metastore.
> 18/02/12 11:23:05 INFO hive.metastore: Closed a connection to metastore, current connections: 0
> 18/02/12 11:23:05 ERROR core.Locomotive: Failed to replicate: karmalab-dev:ddelcastilloperez.access_log to aws-data-lab:bdp.access_log_dev.
> com.hotels.bdp.circustrain.api.metastore.MetaStoreClientException: Cannot fetch table metadata for 'ddelcastilloperez.access_log'
>     at com.hotels.bdp.circustrain.core.HiveEndpoint.getTableAndStatistics(HiveEndpoint.java:132)
>     at com.hotels.bdp.circustrain.core.source.Source.getTableAndStatistics(Source.java:111)
>     at com.hotels.bdp.circustrain.core.ReplicationFactoryImpl.validate(ReplicationFactoryImpl.java:170)
>     at com.hotels.bdp.circustrain.core.ReplicationFactoryImpl.newInstance(ReplicationFactoryImpl.java:85)
>     at com.hotels.bdp.circustrain.core.Locomotive.run(Locomotive.java:111)
>     at org.springframework.boot.SpringApplication.callRunner(SpringApplication.java:791)
>     at org.springframework.boot.SpringApplication.callRunners(SpringApplication.java:781)
>     at org.springframework.boot.SpringApplication.afterRefresh(SpringApplication.java:771)
>     at org.springframework.boot.SpringApplication.run(SpringApplication.java:307)
>     at com.hotels.bdp.circustrain.CircusTrain.main(CircusTrain.java:97)
>     at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
>     at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
>     at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
>     at java.lang.reflect.Method.invoke(Method.java:498)
>     at org.apache.hadoop.util.RunJar.run(RunJar.java:221)
>     at org.apache.hadoop.util.RunJar.main(RunJar.java:136)
> Caused by: org.apache.thrift.TApplicationException: Invalid method name: 'get_table_req'
>     at org.apache.thrift.TApplicationException.read(TApplicationException.java:111)
>     at org.apache.thrift.TServiceClient.receiveBase(TServiceClient.java:79)
>     at org.apache.hadoop.hive.metastore.api.ThriftHiveMetastore$Client.recv_get_table_req(ThriftHiveMetastore.java:1563)
>     at org.apache.hadoop.hive.metastore.api.ThriftHiveMetastore$Client.get_table_req(ThriftHiveMetastore.java:1550)
>     at org.apache.hadoop.hive.metastore.HiveMetaStoreClient.getTable(HiveMetaStoreClient.java:1344)
>     at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
>     at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
>     at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
>     at java.lang.reflect.Method.invoke(Method.java:498)
>     at org.apache.hadoop.hive.metastore.RetryingMetaStoreClient.invoke(RetryingMetaStoreClient.java:169)
>     at com.sun.proxy.$Proxy91.getTable(Unknown Source)
>     at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
>     at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
>     at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
>     at java.lang.reflect.Method.invoke(Method.java:498)
>     at com.hotels.bdp.circustrain.api.metastore.CloseableMetaStoreClientFactory$CloseableMetaStoreClientInvocationHandler.invoke(CloseableMetaStoreClientFactory.java:47)
>     at com.sun.proxy.$Proxy92.getTable(Unknown Source)
>     at com.hotels.bdp.circustrain.core.HiveEndpoint.getTableAndStatistics(HiveEndpoint.java:114)
>     ... 15 more
> 18/02/12 11:23:05 INFO metrics.MetricSender: bdp.access_log_dev.replication_time : -1
> 18/02/12 11:23:05 INFO metrics.MetricSender: bdp.access_log_dev.completion_code : -1
> 18/02/12 11:23:05 INFO metrics.MetricSender: bdp.access_log_dev.bytes_replicated : 0
> 18/02/12 11:23:05 ERROR event.LoggingListener:REPLICATION_EVENTS: [event-id-unavailable] Failed to replicate 'karmalab-dev:ddelcastilloperez.access_log' to 'aws-data-lab:bdp.access_log_dev' with error 'Cannot fetch table metadata for 'ddelcastilloperez.access_log''
> 18/02/12 11:23:05 INFO metrics.MetricSender: tables_replicated : 1
> 18/02/12 11:23:05 INFO metrics.MetricSender: completion_code : -1
> 18/02/12 11:23:05 INFO housekeeping.HousekeepingRunner: Housekeeping at instant 2018-02-09T11:23:05.103Z has started
> 18/02/12 11:23:05 INFO housekeeping.HousekeepingRunner: Housekeeping at instant 2018-02-09T11:23:05.103Z has finished
> 18/02/12 11:23:05 INFO metrics.MetricSender: housekeeping : 1
> 18/02/12 11:23:05 INFO circustrain.CircusTrain: Started CircusTrain in 11.598 seconds (JVM running for 13.477)
> 18/02/12 11:23:05 INFO annotation.AnnotationConfigApplicationContext: Closing org.springframework.context.annotation.AnnotationConfigApplicationContext@fb5aeed: startup date [Mon Feb 12 11:22:54 UTC 2018]; root of context hierarchy
> 18/02/12 11:23:05 INFO annotation.AnnotationMBeanExporter: Unregistering JMX-exposed beans on shutdown
> 18/02/12 11:23:05 INFO annotation.AnnotationMBeanExporter: Unregistering JMX-exposed beans
> 18/02/12 11:23:05 INFO jpa.LocalContainerEntityManagerFactoryBean: Closing JPA EntityManagerFactory for persistence unit 'default'
> 18/02/12 11:23:05 INFO hikari.HikariDataSource: HikariPool-1 - Shutdown initiated...
> 18/02/12 11:23:05 INFO pool.HikariPool: HikariPool-1 - Close initiated...
> 18/02/12 11:23:05 INFO pool.HikariPool: HikariPool-1 - Closed.
> 18/02/12 11:23:05 INFO hikari.HikariDataSource: HikariPool-1 - Shutdown completed.

--- a/circus-train-api/pom.xml
+++ b/circus-train-api/pom.xml
@@ -34,7 +34,7 @@
       <groupId>io.dropwizard.metrics</groupId>
       <artifactId>metrics-core</artifactId>
     </dependency>
-    
+
     <!-- Test -->
     <dependency>
       <groupId>org.hamcrest</groupId>

--- a/circus-train-api/pom.xml
+++ b/circus-train-api/pom.xml
@@ -37,18 +37,24 @@
 
     <!-- Test -->
     <dependency>
+      <groupId>com.hotels</groupId>
+      <artifactId>beeju</artifactId>
+    </dependency>
+    <dependency>
       <groupId>org.hamcrest</groupId>
       <artifactId>hamcrest-library</artifactId>
-      <scope>test</scope>
     </dependency>
     <dependency>
       <groupId>junit</groupId>
       <artifactId>junit</artifactId>
-      <scope>test</scope>
     </dependency>
     <dependency>
       <groupId>org.mockito</groupId>
       <artifactId>mockito-core</artifactId>
+    </dependency>
+    <dependency>
+      <groupId>org.apache.hive</groupId>
+      <artifactId>hive-exec</artifactId>
       <scope>test</scope>
     </dependency>
   </dependencies>

--- a/circus-train-api/src/main/java/com/hotels/bdp/circustrain/api/metastore/CloseableMetaStoreClientFactory.java
+++ b/circus-train-api/src/main/java/com/hotels/bdp/circustrain/api/metastore/CloseableMetaStoreClientFactory.java
@@ -25,6 +25,9 @@ import org.apache.thrift.TApplicationException;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
+import com.google.common.annotations.VisibleForTesting;
+
+import com.hotels.bdp.circustrain.api.metastore.compatibility.HiveMetaStoreClientCompatibility;
 import com.hotels.bdp.circustrain.api.metastore.compatibility.HiveMetaStoreClientCompatibility12x;
 
 public final class CloseableMetaStoreClientFactory {
@@ -33,24 +36,36 @@ public final class CloseableMetaStoreClientFactory {
   private CloseableMetaStoreClientFactory() {}
 
   public static CloseableMetaStoreClient newInstance(IMetaStoreClient delegate) {
+    HiveMetaStoreClientCompatibility compatibility = null;
+    try {
+      compatibility = new HiveMetaStoreClientCompatibility12x(delegate);
+    } catch (Throwable t) {
+      log.warn("Unable to initialize compatibility", t);
+    }
+    return newInstance(delegate, compatibility);
+  }
+
+  @VisibleForTesting
+  static CloseableMetaStoreClient newInstance(
+      IMetaStoreClient delegate,
+      HiveMetaStoreClientCompatibility compatibility) {
     ClassLoader classLoader = CloseableMetaStoreClient.class.getClassLoader();
     Class<?>[] interfaces = new Class<?>[] { CloseableMetaStoreClient.class };
-    CloseableMetaStoreClientInvocationHandler handler = new CloseableMetaStoreClientInvocationHandler(delegate);
+    CloseableMetaStoreClientInvocationHandler handler = new CloseableMetaStoreClientInvocationHandler(delegate,
+        compatibility);
     return (CloseableMetaStoreClient) Proxy.newProxyInstance(classLoader, interfaces, handler);
   }
 
   static class CloseableMetaStoreClientInvocationHandler implements InvocationHandler {
 
     private final IMetaStoreClient delegate;
-    private HiveMetaStoreClientCompatibility12x compatibility;
+    private final HiveMetaStoreClientCompatibility compatibility;
 
-    CloseableMetaStoreClientInvocationHandler(IMetaStoreClient delegate) {
+    CloseableMetaStoreClientInvocationHandler(
+        IMetaStoreClient delegate,
+        HiveMetaStoreClientCompatibility compatibility) {
       this.delegate = delegate;
-      try {
-        compatibility = new HiveMetaStoreClientCompatibility12x(delegate);
-      } catch (Throwable t) {
-        log.warn("Unable to initialize compatibility", t);
-      }
+      this.compatibility = compatibility;
     }
 
     @Override

--- a/circus-train-api/src/main/java/com/hotels/bdp/circustrain/api/metastore/CloseableMetaStoreClientFactory.java
+++ b/circus-train-api/src/main/java/com/hotels/bdp/circustrain/api/metastore/CloseableMetaStoreClientFactory.java
@@ -1,5 +1,5 @@
 /**
- * Copyright (C) 2016-2017 Expedia Inc.
+ * Copyright (C) 2016-2018 Expedia Inc.
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
@@ -21,8 +21,14 @@ import java.lang.reflect.Method;
 import java.lang.reflect.Proxy;
 
 import org.apache.hadoop.hive.metastore.IMetaStoreClient;
+import org.apache.thrift.TApplicationException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.hotels.bdp.circustrain.api.metastore.compatibility.HiveMetaStoreClientCompatibility12x;
 
 public final class CloseableMetaStoreClientFactory {
+  private static final Logger log = LoggerFactory.getLogger(CloseableMetaStoreClientFactory.class);
 
   private CloseableMetaStoreClientFactory() {}
 
@@ -36,9 +42,15 @@ public final class CloseableMetaStoreClientFactory {
   static class CloseableMetaStoreClientInvocationHandler implements InvocationHandler {
 
     private final IMetaStoreClient delegate;
+    private HiveMetaStoreClientCompatibility12x compatibility;
 
     CloseableMetaStoreClientInvocationHandler(IMetaStoreClient delegate) {
       this.delegate = delegate;
+      try {
+        compatibility = new HiveMetaStoreClientCompatibility12x(delegate);
+      } catch (Throwable t) {
+        log.warn("Unable to initialize compatibility", t);
+      }
     }
 
     @Override
@@ -46,8 +58,30 @@ public final class CloseableMetaStoreClientFactory {
       try {
         return method.invoke(delegate, args);
       } catch (InvocationTargetException e) {
+        try {
+          if (compatibility != null && e.getCause().getClass().isAssignableFrom(TApplicationException.class)) {
+            return invokeCompatibility(method, args);
+          }
+        } catch (Throwable t) {
+          log.warn("Unable to run compatibility for metastore client method {}. Will rethrow original exception: ",
+              method.getName(), t);
+        }
         throw e.getCause();
       }
+    }
+
+    private Object invokeCompatibility(Method method, Object[] args) throws Throwable {
+      Class<?>[] argTypes = getTypes(args);
+      Method compatibilityMethod = compatibility.getClass().getMethod(method.getName(), argTypes);
+      return compatibilityMethod.invoke(compatibility, args);
+    }
+
+    private static Class<?>[] getTypes(Object[] args) {
+      Class<?>[] argTypes = new Class<?>[args.length];
+      for (int i = 0; i < args.length; ++i) {
+        argTypes[i] = args[i].getClass();
+      }
+      return argTypes;
     }
 
   }

--- a/circus-train-api/src/main/java/com/hotels/bdp/circustrain/api/metastore/compatibility/HiveMetaStoreClientCompatibility.java
+++ b/circus-train-api/src/main/java/com/hotels/bdp/circustrain/api/metastore/compatibility/HiveMetaStoreClientCompatibility.java
@@ -1,0 +1,27 @@
+/**
+ * Copyright (C) 2016-2018 Expedia Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.hotels.bdp.circustrain.api.metastore.compatibility;
+
+import org.apache.hadoop.hive.metastore.api.MetaException;
+import org.apache.hadoop.hive.metastore.api.NoSuchObjectException;
+import org.apache.hadoop.hive.metastore.api.Table;
+import org.apache.thrift.TException;
+
+public interface HiveMetaStoreClientCompatibility {
+
+  Table getTable(String dbname, String name) throws MetaException, TException, NoSuchObjectException;
+
+}

--- a/circus-train-api/src/main/java/com/hotels/bdp/circustrain/api/metastore/compatibility/HiveMetaStoreClientCompatibility12x.java
+++ b/circus-train-api/src/main/java/com/hotels/bdp/circustrain/api/metastore/compatibility/HiveMetaStoreClientCompatibility12x.java
@@ -43,7 +43,7 @@ public class HiveMetaStoreClientCompatibility12x implements HiveMetaStoreClientC
         client = getField(handler, "base");
         continue;
       }
-      // Others handlers can be added here
+      // Other handlers can be added here
       throw new RuntimeException("Unknown InvocationHandler " + handler.getClass());
     }
     tServiceClient = getField(client, "client");

--- a/circus-train-api/src/main/java/com/hotels/bdp/circustrain/api/metastore/compatibility/HiveMetaStoreClientCompatibility12x.java
+++ b/circus-train-api/src/main/java/com/hotels/bdp/circustrain/api/metastore/compatibility/HiveMetaStoreClientCompatibility12x.java
@@ -1,5 +1,5 @@
 /**
- * Copyright (C) 2016-2018 Expedia Inc.
+ * Copyright (C) 2016-2018 Expedia Inc and the original Apache hive-metastore contributors.
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
@@ -32,7 +32,7 @@ import org.apache.thrift.TBase;
 import org.apache.thrift.TException;
 import org.apache.thrift.TServiceClient;
 
-public class HiveMetaStoreClientCompatibility12x {
+public class HiveMetaStoreClientCompatibility12x implements HiveMetaStoreClientCompatibility {
 
   private final TServiceClient tServiceClient;
 
@@ -43,7 +43,7 @@ public class HiveMetaStoreClientCompatibility12x {
         client = getField(handler, "base");
         continue;
       }
-      // TODO add CloseableMetaStoreClientInvocationHandler ?
+      // Others handlers can be added here
       throw new RuntimeException("Unknown InvocationHandler " + handler.getClass());
     }
     tServiceClient = getField(client, "client");
@@ -94,15 +94,30 @@ public class HiveMetaStoreClientCompatibility12x {
     }
   }
 
+  /*
+   * Based on https://github.com/apache/hive/blob/release-1.2.1/metastore/src/java/org/apache/hadoop/hive/metastore/
+   * HiveMetaStoreClient.java#L1206
+   */
+  @Override
   public Table getTable(String dbname, String name) throws MetaException, TException, NoSuchObjectException {
     return deepCopy(get_table(dbname, name));
   }
 
+  /*
+   * Copied from Hive 1.2.1 ThriftHiveMetastore.Client#get_table(String,String) - see
+   * https://raw.githubusercontent.com/apache/hive/release-1.2.1/metastore/src/gen/thrift/gen-javabean/org/apache/hadoop
+   * /hive/metastore/api/ThriftHiveMetastore.java
+   */
   private Table get_table(String dbname, String tbl_name) throws MetaException, NoSuchObjectException, TException {
     send_get_table(dbname, tbl_name);
     return recv_get_table();
   }
 
+  /*
+   * Copied from Hive 1.2.1 ThriftHiveMetastore.Client#send_get_table(String,String) - see
+   * https://raw.githubusercontent.com/apache/hive/release-1.2.1/metastore/src/gen/thrift/gen-javabean/org/apache/hadoop
+   * /hive/metastore/api/ThriftHiveMetastore.java
+   */
   private void send_get_table(String dbname, String tbl_name) throws TException {
     ThriftHiveMetastore.get_table_args args = new ThriftHiveMetastore.get_table_args();
     args.setDbname(dbname);
@@ -110,6 +125,11 @@ public class HiveMetaStoreClientCompatibility12x {
     sendBase("get_table", args);
   }
 
+  /*
+   * Based on Hive 1.2.1 ThriftHiveMetastore.Client#recv_get_table() - see
+   * https://raw.githubusercontent.com/apache/hive/release-1.2.1/metastore/src/gen/thrift/gen-javabean/org/apache/hadoop
+   * /hive/metastore/api/ThriftHiveMetastore.java
+   */
   private Table recv_get_table() throws MetaException, NoSuchObjectException, TException {
     ThriftHiveMetastore.get_table_result result = new ThriftHiveMetastore.get_table_result();
     receiveBase(result, "get_table");

--- a/circus-train-api/src/main/java/com/hotels/bdp/circustrain/api/metastore/compatibility/HiveMetaStoreClientCompatibility12x.java
+++ b/circus-train-api/src/main/java/com/hotels/bdp/circustrain/api/metastore/compatibility/HiveMetaStoreClientCompatibility12x.java
@@ -1,0 +1,128 @@
+/**
+ * Copyright (C) 2016-2018 Expedia Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.hotels.bdp.circustrain.api.metastore.compatibility;
+
+import java.lang.reflect.Field;
+import java.lang.reflect.InvocationHandler;
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
+import java.lang.reflect.Proxy;
+
+import org.apache.hadoop.hive.metastore.IMetaStoreClient;
+import org.apache.hadoop.hive.metastore.RetryingMetaStoreClient;
+import org.apache.hadoop.hive.metastore.api.MetaException;
+import org.apache.hadoop.hive.metastore.api.NoSuchObjectException;
+import org.apache.hadoop.hive.metastore.api.Table;
+import org.apache.hadoop.hive.metastore.api.ThriftHiveMetastore;
+import org.apache.thrift.TApplicationException;
+import org.apache.thrift.TBase;
+import org.apache.thrift.TException;
+import org.apache.thrift.TServiceClient;
+
+public class HiveMetaStoreClientCompatibility12x {
+
+  private final TServiceClient tServiceClient;
+
+  public HiveMetaStoreClientCompatibility12x(IMetaStoreClient client) {
+    while (Proxy.isProxyClass(client.getClass())) {
+      InvocationHandler handler = Proxy.getInvocationHandler(client);
+      if (handler.getClass().isAssignableFrom(RetryingMetaStoreClient.class)) {
+        client = getField(handler, "base");
+        continue;
+      }
+      // TODO add CloseableMetaStoreClientInvocationHandler ?
+      throw new RuntimeException("Unknown InvocationHandler " + handler.getClass());
+    }
+    tServiceClient = getField(client, "client");
+  }
+
+  private static <T> T getField(Object object, String fieldName) {
+    try {
+      Field field = object.getClass().getDeclaredField(fieldName);
+      T result = null;
+      if (field.isAccessible()) {
+        result = (T) field.get(object);
+      } else {
+        field.setAccessible(true);
+        result = (T) field.get(object);
+        field.setAccessible(false);
+      }
+      return result;
+    } catch (Exception e) {
+      throw new RuntimeException("Unable to hack client", e);
+    }
+  }
+
+  private static Table deepCopy(Table table) {
+    return table.deepCopy();
+  }
+
+  private void sendBase(String methodName, TBase<?, ?> args) throws TException {
+    try {
+      Method sendBase = TServiceClient.class.getDeclaredMethod("sendBase", String.class, TBase.class);
+      sendBase.setAccessible(true);
+      sendBase.invoke(tServiceClient, methodName, args);
+      sendBase.setAccessible(false);
+    } catch (NoSuchMethodException | SecurityException | IllegalAccessException | IllegalArgumentException
+        | InvocationTargetException e) {
+      throw new RuntimeException("Unable to hack sendBase", e);
+    }
+  }
+
+  private void receiveBase(TBase<?, ?> result, String methodName) throws TException {
+    try {
+      Method receiveBase = TServiceClient.class.getDeclaredMethod("receiveBase", TBase.class, String.class);
+      receiveBase.setAccessible(true);
+      receiveBase.invoke(tServiceClient, result, methodName);
+      receiveBase.setAccessible(false);
+    } catch (NoSuchMethodException | SecurityException | IllegalAccessException | IllegalArgumentException
+        | InvocationTargetException e) {
+      throw new RuntimeException("Unable to hack receiveBase", e);
+    }
+  }
+
+  public Table getTable(String dbname, String name) throws MetaException, TException, NoSuchObjectException {
+    return deepCopy(get_table(dbname, name));
+  }
+
+  private Table get_table(String dbname, String tbl_name) throws MetaException, NoSuchObjectException, TException {
+    send_get_table(dbname, tbl_name);
+    return recv_get_table();
+  }
+
+  private void send_get_table(String dbname, String tbl_name) throws TException {
+    ThriftHiveMetastore.get_table_args args = new ThriftHiveMetastore.get_table_args();
+    args.setDbname(dbname);
+    args.setTbl_name(tbl_name);
+    sendBase("get_table", args);
+  }
+
+  private Table recv_get_table() throws MetaException, NoSuchObjectException, TException {
+    ThriftHiveMetastore.get_table_result result = new ThriftHiveMetastore.get_table_result();
+    receiveBase(result, "get_table");
+    if (result.isSetSuccess()) {
+      return result.getSuccess();
+    }
+    if (result.getO1() != null) {
+      throw result.getO1();
+    }
+    if (result.getO2() != null) {
+      throw result.getO2();
+    }
+    throw new TApplicationException(5, "get_table failed: unknown result");
+  }
+
+}

--- a/circus-train-api/src/test/java/com/hotels/bdp/circustrain/api/metastore/CloseableMetaStoreClientFactoryTest.java
+++ b/circus-train-api/src/test/java/com/hotels/bdp/circustrain/api/metastore/CloseableMetaStoreClientFactoryTest.java
@@ -1,5 +1,5 @@
 /**
- * Copyright (C) 2016-2017 Expedia Inc.
+ * Copyright (C) 2016-2018 Expedia Inc.
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
@@ -16,31 +16,40 @@
 package com.hotels.bdp.circustrain.api.metastore;
 
 import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.when;
 
 import org.apache.hadoop.hive.metastore.IMetaStoreClient;
+import org.apache.thrift.TApplicationException;
 import org.apache.thrift.TException;
 import org.junit.Test;
 import org.junit.runner.RunWith;
 import org.mockito.Mock;
 import org.mockito.runners.MockitoJUnitRunner;
 
-import com.hotels.bdp.circustrain.api.metastore.CloseableMetaStoreClient;
-import com.hotels.bdp.circustrain.api.metastore.CloseableMetaStoreClientFactory;
+import com.hotels.bdp.circustrain.api.metastore.compatibility.HiveMetaStoreClientCompatibility;
 
 @RunWith(MockitoJUnitRunner.class)
 public class CloseableMetaStoreClientFactoryTest {
 
-  @Mock
-  private IMetaStoreClient delegate;
+  private @Mock IMetaStoreClient delegate;
+  private @Mock HiveMetaStoreClientCompatibility compatibility;
 
   @Test
   public void typical() throws TException {
     try (CloseableMetaStoreClient wrapped = CloseableMetaStoreClientFactory.newInstance(delegate)) {
       wrapped.unlock(1L);
     }
-
     verify(delegate).unlock(1L);
     verify(delegate).close();
+  }
+
+  @Test
+  public void compatibility() throws TException {
+    when(delegate.getTable("db", "tbl")).thenThrow(new TApplicationException());
+    try (CloseableMetaStoreClient wrapped = CloseableMetaStoreClientFactory.newInstance(delegate, compatibility)) {
+      wrapped.getTable("db", "tbl");
+    }
+    verify(compatibility).getTable("db", "tbl");
   }
 
 }

--- a/circus-train-api/src/test/java/com/hotels/bdp/circustrain/api/metastore/compatibility/HiveMetaStoreClientCompatibility12xTest.java
+++ b/circus-train-api/src/test/java/com/hotels/bdp/circustrain/api/metastore/compatibility/HiveMetaStoreClientCompatibility12xTest.java
@@ -1,0 +1,152 @@
+/**
+ * Copyright (C) 2016-2018 Expedia Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.hotels.bdp.circustrain.api.metastore.compatibility;
+
+import static org.mockito.Matchers.eq;
+import static org.mockito.Mockito.spy;
+import static org.mockito.Mockito.verify;
+
+import java.lang.reflect.Field;
+import java.lang.reflect.InvocationHandler;
+import java.lang.reflect.Method;
+import java.lang.reflect.Proxy;
+
+import org.apache.hadoop.hive.metastore.HiveMetaHook;
+import org.apache.hadoop.hive.metastore.HiveMetaHookLoader;
+import org.apache.hadoop.hive.metastore.HiveMetaStoreClient;
+import org.apache.hadoop.hive.metastore.IMetaStoreClient;
+import org.apache.hadoop.hive.metastore.RetryingMetaStoreClient;
+import org.apache.hadoop.hive.metastore.api.FieldSchema;
+import org.apache.hadoop.hive.metastore.api.GetTableRequest;
+import org.apache.hadoop.hive.metastore.api.GetTableResult;
+import org.apache.hadoop.hive.metastore.api.MetaException;
+import org.apache.hadoop.hive.metastore.api.NoSuchObjectException;
+import org.apache.hadoop.hive.metastore.api.SerDeInfo;
+import org.apache.hadoop.hive.metastore.api.StorageDescriptor;
+import org.apache.hadoop.hive.metastore.api.Table;
+import org.apache.hadoop.hive.metastore.api.ThriftHiveMetastore;
+import org.apache.hadoop.hive.metastore.api.ThriftHiveMetastore.get_table_args;
+import org.apache.thrift.TApplicationException;
+import org.apache.thrift.TBase;
+import org.apache.thrift.TException;
+import org.apache.thrift.TServiceClient;
+import org.junit.Before;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.mockito.runners.MockitoJUnitRunner;
+
+import com.google.common.collect.ImmutableList;
+
+import com.hotels.beeju.ThriftHiveMetaStoreJUnitRule;
+
+@RunWith(MockitoJUnitRunner.class)
+public class HiveMetaStoreClientCompatibility12xTest {
+
+  private static final String DATABASE = "db";
+  private static final String TABLE = "tbl";
+
+  private static class RelaxedTServiceClient extends ThriftHiveMetastore.Client {
+    private final ThriftHiveMetastore.Client delegate;
+
+    public RelaxedTServiceClient(ThriftHiveMetastore.Client delegate) {
+      super(delegate.getInputProtocol());
+      this.delegate = delegate;
+    }
+
+    @Override
+    public GetTableResult get_table_req(GetTableRequest req) throws MetaException, NoSuchObjectException, TException {
+      // Safeguard: this method should not be invoked since we are testing directrly calling the compatibility class
+      throw new TApplicationException();
+    }
+
+    @Override
+    protected void sendBase(String methodName, TBase<?, ?> args) throws TException {
+      try {
+        Method sendBase = TServiceClient.class.getDeclaredMethod("sendBase", String.class, TBase.class);
+        sendBase.setAccessible(true);
+        sendBase.invoke(delegate, methodName, args);
+      } catch (Exception e) {
+        throw new TException(e);
+      }
+    }
+
+    @Override
+    protected void receiveBase(TBase<?, ?> result, String methodName) throws TException {
+      try {
+        Method sendBase = TServiceClient.class.getDeclaredMethod("receiveBase", TBase.class, String.class);
+        sendBase.setAccessible(true);
+        sendBase.invoke(delegate, result, methodName);
+      } catch (Exception e) {
+        throw new TException(e);
+      }
+    }
+  }
+
+  public final @Rule ThriftHiveMetaStoreJUnitRule metastore = new ThriftHiveMetaStoreJUnitRule(DATABASE);
+
+  @Before
+  public void init() throws Exception {
+    Table table = new Table();
+    table.setDbName(DATABASE);
+    table.setTableName(TABLE);
+    table.setSd(new StorageDescriptor());
+    table.getSd().setCols(ImmutableList.of(new FieldSchema("col", "string", null)));
+    table.getSd().setSerdeInfo(new SerDeInfo());
+    metastore.client().createTable(table);
+  }
+
+  private RelaxedTServiceClient prepareClient(IMetaStoreClient metastoreClient) throws Exception {
+    Field field = metastoreClient.getClass().getDeclaredField("client");
+    field.setAccessible(true);
+    ThriftHiveMetastore.Client client = (ThriftHiveMetastore.Client) field.get(metastoreClient);
+    RelaxedTServiceClient relaxedTServiceClient = spy(new RelaxedTServiceClient(client));
+    field.set(metastoreClient, relaxedTServiceClient);
+    return relaxedTServiceClient;
+  }
+
+  private RelaxedTServiceClient prepareRetryingClient(IMetaStoreClient metastoreClient) throws Exception {
+    InvocationHandler handler = Proxy.getInvocationHandler(metastoreClient);
+    Field field = handler.getClass().getDeclaredField("base");
+    field.setAccessible(true);
+    IMetaStoreClient base = (IMetaStoreClient) field.get(handler);
+    return prepareClient(base);
+  }
+
+  @Test
+  public void properClient() throws Exception {
+    IMetaStoreClient metastoreClient = metastore.client();
+    RelaxedTServiceClient tServiceClient = prepareClient(metastoreClient);
+    HiveMetaStoreClientCompatibility12x compatibility = new HiveMetaStoreClientCompatibility12x(metastoreClient);
+    compatibility.getTable(DATABASE, TABLE);
+    verify(tServiceClient).sendBase(eq("get_table"), eq(new get_table_args(DATABASE, TABLE)));
+  }
+
+  @Test
+  public void prepareRetryingClient() throws Exception {
+    IMetaStoreClient metastoreClient = RetryingMetaStoreClient.getProxy(metastore.conf(), new HiveMetaHookLoader() {
+      @Override
+      public HiveMetaHook getHook(Table tbl) throws MetaException {
+        return null;
+      }
+    }, HiveMetaStoreClient.class.getName());
+    RelaxedTServiceClient tServiceClient = prepareRetryingClient(metastoreClient);
+    HiveMetaStoreClientCompatibility12x compatibility = new HiveMetaStoreClientCompatibility12x(metastoreClient);
+    compatibility.getTable(DATABASE, TABLE);
+    verify(tServiceClient).sendBase(eq("get_table"), eq(new get_table_args(DATABASE, TABLE)));
+  }
+
+}

--- a/circus-train-api/src/test/java/com/hotels/bdp/circustrain/api/metastore/compatibility/HiveMetaStoreClientCompatibility12xTest.java
+++ b/circus-train-api/src/test/java/com/hotels/bdp/circustrain/api/metastore/compatibility/HiveMetaStoreClientCompatibility12xTest.java
@@ -69,7 +69,7 @@ public class HiveMetaStoreClientCompatibility12xTest {
 
     @Override
     public GetTableResult get_table_req(GetTableRequest req) throws MetaException, NoSuchObjectException, TException {
-      // Safeguard: this method should not be invoked since we are testing directrly calling the compatibility class
+      // Safeguard: this method should not be invoked since we are testing directly calling the compatibility class
       throw new TApplicationException();
     }
 

